### PR TITLE
Ryanunderhill/c api 8

### PR DESCRIPTION
**Pull request by RyanUnderhill**
This removes reference counting from OrtAllocator. Now a user of the C API just needs to ensure their own allocator outlives all of the objects used by Ort.

If a user uses one of our provided allocators, they just use 'CreateDefaultAllocator' and 'OrtReleaseAllocator' to free it.

### CODE REVIEW

**Comment by pranavsharma** [comment]
@RyanUnderhill once this PR is merged, please merge this with the aiinfra/vsts repo as the automatic merge will fail. Thanks.

**Comment by snnn** [comment]
Also, it removes reference counting from OrtEnv. Please add a note in your comments when completing this PR.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -183,7 +180,7 @@ typedef void(ORT_API_CALL* OrtLoggingFunction)(
 
 /**
  * OrtEnv is process-wide. For each process, only one OrtEnv can be created.
- * \param out Should be freed by `OrtReleaseObject` after use
+ * \param out Should be freed by `OrtReleaseEnv` after use
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Oops, I changed the c# part before I fixed this. Fixed now.

File: onnxruntime/core/session/default_cpu_allocator_c_api.cc [hunk 2]
```diff
@@ -5,81 +5,55 @@
...
-void Free(void* p) {
-  return ::free(p);
-}
-const OrtAllocatorInfo* Info() const {
-  return cpuAllocatorInfo;
-}
-ORT_ALLOCATOR_IMPL_END
+ private:
+  OrtDefaultAllocator(const OrtDefaultAllocator&) = delete;
+  OrtDefaultAllocator& operator=(const OrtDefaultAllocator&) = delete;
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply onnxruntime/core/session/default_cpu_allocator_c_api.cc]: I kept that part unchanged, when I tried that macro it didn't work. Apparently it's not including common.h from this file. Still want me to change it?

[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/core/session/default_cpu_allocator_c_api.cc]: NO. Thanks.

**Pull request merged**